soci/gzip_indexer.go

```diff
@@ -0,0 +1,147 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package soci
+
+// #cgo CFLAGS: -I${SRCDIR}/../c/
+// #cgo LDFLAGS: -L${SRCDIR}/../out -lindexer -lz
+// #include "indexer.h"
+// #include <stdlib.h>
+// #include <stdint.h>
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type GzipIndexer struct {
+	index *C.struct_gzip_index
+}
+
+// GenerateIndex wraps `C.generate_index` and should be used for index generation instead of relying on C code.
+// Returns the byte slice containing the `gzip_index`.
+func GenerateIndex(gzipFile string, spanSize int64) ([]byte, error) {
+	cstr := C.CString(gzipFile)
+	defer C.free(unsafe.Pointer(cstr))
+
+	var index *C.struct_gzip_index
+	ret := C.generate_index(cstr, C.off_t(spanSize), &index)
+	if int(ret) < 0 {
+		return nil, fmt.Errorf("could not generate gzip index. gzip error: %v", ret)
+	}
+	defer C.free(unsafe.Pointer(index))
+
+	blobSize := C.get_blob_size(index)
+	bytes := make([]byte, uint64(blobSize))
+	if bytes == nil {
+		return nil, fmt.Errorf("could not allocate byte array of size %d", blobSize)
+	}
+
+	ret = C.index_to_blob(index, unsafe.Pointer(&bytes[0]))
+	if int(ret) <= 0 {
+		return nil, fmt.Errorf("could not serialize gzip index to byte array; gzip error: %v", ret)
+	}
+
+	return bytes, nil
+}
+
+// Close calls `C.free` on the pointer to `C.struct_gzip_index`.
+func (i *GzipIndexer) Close() {
+	if i.index != nil {
+		C.free(unsafe.Pointer(i.index))
+	}
+}
+
+// GetMaxSpanID returns the max span ID.
+func (i *GzipIndexer) GetMaxSpanID() SpanID {
+	return SpanID(i.index.have - 1)
+}
+
+// GetSpanIDByUncompressedOffset returns the ID of the span containing the data pointed by uncompressed offset.
+func (i *GzipIndexer) GetSpanIDByUncompressedOffset(offset FileSize) SpanID {
```

Comment:
Naming nit: "For" not "By". `GetSpanIDForUncompressedOffset`. Alternatively, I would avoid "Get" in all of these function names, since they are converters, not getters. So
* UncompressedOffsetToSpanID
* SpanIDToCompressedOffset
* SpanIDToUncompressedOffset

Reply:
Done.